Add a derive macro

DusterTheFirst opened the pull request:
This PR adds a derive macro for the `Lerp` trait that can work on any struct with homogeneous float fields, either all f64 or all f32. They cannot mix.

The derive macro is behind a feature flag as not all users will need the derive macro and it will save them from downloading the separate derive crate.

The derive implementation simply will lerp each of the fields of the struct individually.

DusterTheFirst committed:
feat: Derive macro

DusterTheFirst committed:
fix clippy lints

DusterTheFirst committed:
readme

DusterTheFirst committed:
More readme

coriolinus commented on lerp_derive/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "lerp_derive"
+version = "0.1.0"
+authors = ["coriolinus <coriolinus@gmail.com>"]
```
You may as well add yourself as an author, at least for this derive crate.

coriolinus commented on lerp_derive/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+keywords = ["lerp", "math", "interpolate", "extrapolate", "numeric"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Get rid of this comment please.

coriolinus reviewed (commented):
Wow, impressive! I don't have time to look properly into this at the moment, but I'm going to try to get this sorted by the end of the week. Thanks for contributing this!

DusterTheFirst committed:
fix Cargo.toml

DusterTheFirst committed:
Documentation and refactor

The derive macro has been refactored to use results rather than quote_spanned which is a bit more neat

coriolinus commented on lerp_derive/src/derive.rs:
```diff
@@ -0,0 +1,61 @@
+                Ok(quote! {
+                    #[automatically_derived]
+                    impl Lerp<#ty> for #name {
+                        fn lerp(self, other: Self, t: #ty) -> Self {
```
Extracting `ty` from the first field of the input struct is clever, but restrictive: it sharply limits the structs on which this derive macro can be applied. I think we can be more general without sacrificing accuracy. Something like this (pseudocode, I haven't worked out all the details):

```rust
impl Lerp<f64> for #name {
  fn lerp(self, other: Self, t: f64) -> Self {
    Self {
      // for each #field in #fields, precompute this:
      match #field.ty {
        f64 => #name: self.#name.lerp(other.#name, t),
        f32 => #name: self.#name.lerp(other.#name, t as f32),
        T if T: Lerp<f64> => #name: self.#name.lerp(other.#name, t),
        _ => // compile error
      }
    }
  }
}

// separate impl Lerp<f32> follows, using the same pattern
```

This pattern would allow mixing arbitrary numeric primitives in a struct for which we `derive(Lerp)`, which makes the macro much more powerful. I would prefer that.

Adding duplicate impls for `f64` and `f32` is just a convenience thing, but I can't think of any reason we'd want not to do it.

DusterTheFirst replied:
That makes sense, Ill see what I can do

coriolinus commented on README.md:
```diff
@@ -26,7 +28,38 @@ assert_eq!(vec![3.0, 3.5, 4.0, 4.5], items);
+and the arrays are `T: Add<Output = T> + Mul<F: Float, Output = T>`, it'll just
+work for them as well.
+
+As well as working for individual float values, the crate also provides a derive
```
```suggestion
## Deriving `Lerp`

As well as working for individual float values, the crate also provides a derive
```

coriolinus commented on README.md:
````diff
@@ -26,7 +28,38 @@ assert_eq!(vec![3.0, 3.5, 4.0, 4.5], items);
+
+```toml
+[dependencies]
+lerp = { version = "0.1", features = ["derive"] }
````
```suggestion
lerp = { version = "0.4", features = ["derive"] }
```

Might as well bump this now: current version is `0.3`, and adding a feature like this is easily worth another version bump.

coriolinus reviewed (commented):
First off, I have to apologize: this fell entirely off my radar, and I never got back to you. Sorry about that, and thanks for requesting review to remind me that this PR existed.

There are a few documentation nits which should be trivial to fix, but the larger issue is that I think this implementation as written is somewhat too restrictive. I've suggested a refactor which should reduce the restrictions.

DusterTheFirst committed:
Apply suggestions from code review

Co-authored-by: Peter Goodspeed-Niklaus <coriolinus@users.noreply.github.com>

DusterTheFirst commented:
Tracking comment for changes:
- [x] Split manual tests and derive tests
- [x] Re-export num-traits traits for consumption by the macro and dependent crates
- [x] Create manual mockup tests to layout the output for the derive macro
- [x] Update the derive macro
- [x] Update the documentation for the derive macro
- [x] Tests pass
    - [x] Manual tests
    - [x] Derive tests
    - [ ] ~~Try-build tests?~~

DusterTheFirst committed:
Add manual implementation to layout the derive macro output

DusterTheFirst commented:
@coriolinus I have layed out what I think the derive macro could produce that would be more generic and allow for `Lerp` implementations over both float types, mixing of struct float types, and allowing for any implementor of `Lerp` to be used in the derive macro.

The main focal test is this one: https://github.com/coriolinus/lerp-rs/blob/9c73e97e532436383f3e6a6badad833a96cd59e1/src/lib.rs#L428-L498

~~Ill update the other tests to be generic too~~

The only thing I could see as being a problem with this implementation is the line

https://github.com/coriolinus/lerp-rs/blob/9c73e97e532436383f3e6a6badad833a96cd59e1/src/lib.rs#L452

which assumes the struct derives Lerp generically for all types that implement `Float`. But we may be able to get around that by letting the user put an attribute on the field in the struct specifying what float type to use for the lerp:

```rs
// Assume that InternalData exists, and implements Lerp<f64>
#[derive(PartialEq, Debug, Lerp)]
struct Data {
    #[lerp(f64)]
    a: InternalData,
    b: f32,
}

// And if InternalData exists and implements Lerp<F: Float>
#[derive(PartialEq, Debug, Lerp)]
struct Data {
    a: InternalData,
    b: f32,
}

// Or we could make it mandetory to specify what Lerp implementation is used, so the above would become
#[derive(PartialEq, Debug, Lerp)]
struct Data {
    #[lerp(Float)] // Or something telling it that it is generic
    a: InternalData,
    b: f32,
}
```

Ill wait to hear your opinion on this before updating the macro.

DusterTheFirst committed:
All manual impls are generic

DusterTheFirst committed:
slight syntax changes

coriolinus commented:
> The only thing I could see as being a problem with this implementation is the line ... which assumes the struct derives Lerp generically for all types that implement `Float`

Agree with this concern: if the derived impl is over `F: Float`, then this won't compile:

```rust
struct OldStruct;

/// impl'd this manually back in the day
impl Lerp<f32> for OldStruct { ... }

#[derive(Lerp)]
struct NewStruct {
    old: OldStruct,
}
```

Not a fan of field-level attributes for this case though: for that to work properly, the macro would have to unify the attributes of all fields, and ensure they're all common. Syntax like this seems much simpler:

```rust
use example::BigFloat;

#[derive(Lerp)]
#[impl_lerp_on(f32, BigFloat)]
struct NewStruct {
    old: OldStruct,
}
````

The rule would then be: if `impl_lerp_on` is unspecified, we generic a generic impl on `F: Float`. For all types which contain only primitives, this is the simplest and most powerful option. If it is specified, then we generate a specific impl for each item in the list of items, and no generic impl. Make sense?

DusterTheFirst commented:
> Not a fan of field-level attributes for this case though: for that to work properly, the macro would have to unify the attributes of all fields, and ensure they're all common. 

I don't really understand what you are getting at here. What do you mean by unify the attributes of all fields. I am pretty sure that using the generic `F: Float` implementation, if one of the fields is a struct and derives a specific lerp, such as `Lerp<f32>`, the macro would be able to generate code that would use `num_traits::cast` to cast the t value to the one specified for that field. It would also allow for mixing and matching structures that have different lerp implementations.

The produced code for something like this:
```
// Struct only implementing `Lerp<f64>`
struct InternalData(f64, f32); 

// Some manual implementation of lerp for `f64` only
impl Lerp<f64> for InternalData { 
   fn lerp(self, other: Self, t: f64) -> Self { 
       Self( 
           self.0.lerp(other.0, t), 
           self.1.lerp(other.1, t as f32), 
       ) 
   } 
} 

#[derive(PartialEq, Debug, Lerp)]
struct Data {
    #[lerp(f64)]
    a: InternalData,
    b: f32,
}
```

would look something like this:
```rs
impl<F: crate::Float> Lerp<F> for Data { 
    fn lerp(self, other: Self, t: F) -> Self { 
        Self { 
            a: self.a.lerp(other.a, crate::num_traits::cast::<_, f64>(t).unwrap()), // Allows this struct to Lerp<f64>
            b: self.b.lerp(other.b, crate::num_traits::cast::<_, f32>(t).unwrap()), // While this field Lerps<f32>
        } 
    } 
} 
```

The fields would be able to mix and match what Lerp<> they used

coriolinus commented:
Ah, I see now what you're getting at. I'd initially misunderstood what the generated code would look like given field-level attributes. Your example seems reasonable.

Two more edits to the generated code are necessary, I think:

```diff
- impl<F: crate::Float> Lerp<F> for Data
+ impl<F> Lerp<F> for Data
+   where
+     F: crate::Float + crate::NumCast,
```

I don't think this will compile without a bound on `NumCast`.

```diff
- a: self.a.lerp(other.a, crate::num_traits::cast::<_, f64>(t).unwrap())
+ a: self.a.lerp(other.a, crate::num_traits::cast::<_, f64>(t).expect("it should be safe to cast to f64 for any Float + NumCast")),
```

If this does fail, this should at least improve the panic message.

DusterTheFirst commented:
> If this does fail, this should at least improve the panic message.

Yah, I do plan on making the errors better.

regarding the NumCast trait bound, my manual implementation compiles just fine being generic over just Float, so it may not be necessary

https://github.com/coriolinus/lerp-rs/blob/16a037fcfc8d5592fcac11c070d8fafe89ea57de/src/lib.rs#L400-L425

DusterTheFirst commented:
If we do field level attributes, we can also include a #[lerp(skip)] possibility to just pass the value and not interpolate certain fields

DusterTheFirst committed:
Derive macro implementation

DusterTheFirst commented:
I have modified the derive macro to match what we went over, and in doing so, I realized that we might be able to auto derive lerp for these types too:

       Type::Array(_) => {
            todo!()
        } // TODO: ?

        Type::Slice(_) => {
            todo!()
        } // TODO: ?

        Type::Tuple(_) => {
            todo!()
        } // TODO: ?

DusterTheFirst committed:
many an improvement

- skip or ignore are interchangable
- fix blunder in attribute parsing
- produce dummy implementation when the macro errors out
- more derive tests

coriolinus commented:
Yeah, it's almost certainly a good idea to ensure the macro has the capability to lerp those types. I think it's going to be straightforward for everything except `&'a [F]`, because we can't create a new slice with that lifetime. I think it's fine if we can't derive for that case.

At the same time, having that ability makes it more important to enable `#[lerp(skip)]`.

coriolinus commented on lerp_derive/src/derive.rs:
```diff
@@ -1,60 +1,212 @@
+
+#[derive(Default)]
+struct LerpAttributes {
+    skip: Option<()>,
```
Why `Option<()>` instead of `bool`?

DusterTheFirst replied:
It was left over from a previous idea I had, ill make it a bool

DusterTheFirst replied:
Also, how should we handle the types in arrays? allow for nesting or just 1 level of lerping.

Like should 
```[[f32; 10]; 10]```
be accepted or rejected

coriolinus replied:
Actually, now that I think of it, it'd be simplest to just add a generic

```rust
impl<T, F, N> Lerp<F> for [T; N] { ... }
impl<T, F> Lerp<F> for Vec<T> { ... }
```

and then handle those cases with the generic impl, instead of implementing those cases within the macro.

DusterTheFirst replied:
Yah, looking into it that would be best.

DusterTheFirst replied:
Getting a derive working for arrays is a mess

DusterTheFirst replied:
Adding that derive is a bit hard since the error pops up
```
error[E0119]: conflicting implementations of trait `Lerp<_>` for type `std::vec::Vec<_>`:
   --> src/lib.rs:177:1
    |
166 | / impl<T, F> Lerp<F> for T
167 | | where
168 | |     T: Add<Output = T> + Mul<F, Output = T>,
169 | |     F: Float,
...   |
173 | |     }
174 | | }
    | |_- first implementation here
...
177 |   impl<T: Lerp<F>, F: Float> Lerp<F> for Vec<T> {
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `std::vec::Vec<_>`
    |
    = note: upstream crates may add a new impl of trait `std::ops::Add` for type `std::vec::Vec<_>` in future versions
    = note: downstream crates may implement trait `std::ops::Mul<_>` for type `std::vec::Vec<_>`
    = note: downstream crates may implement trait `std::ops::Mul<_>` for type `std::vec::Vec<_>`
```

This Is hard to get around, without making the previous blanket implementation less general.

coriolinus replied:
Agree. I don't think it's worth stalling this PR any more for that. For now, I think the way forward is to 

- create a new issue in this repo that Lerp derives don't work for vectors or arrays yet
- add a comment in some appropriate place linking to that issue
- just don't handle arrays or vectors in this PR at all

Does that work for you?

DusterTheFirst replied:
Indeed, Ill do that now

DusterTheFirst committed:
UNFINISHED: work towards more lerp derive

DusterTheFirst commented:
I have added an error message pointing to the issue in the macro, and removed all work on arrays.
![image](https://user-images.githubusercontent.com/14093962/105099643-00bf9e80-5a7a-11eb-8bb9-c58b93320dfd.png)

The issue is now #5 and linked both in the issue above and a FIXME comment on the line above.

coriolinus commented on lerp_derive/src/derive.rs:
```diff
@@ -53,54 +44,95 @@ impl Parse for LerpAttributes {
-            if attr.skip.is_some() {
-                return Ok(LerpType::Skip);
+    match ty {
+        Type::Path(TypePath { path, .. }) => Ok(if attr.skip {
```
One final nit, then I think this is done. It looks like we can substantially reduce code duplication by moving the skip check outside the match block, i.e.

```rust
Ok(if attr.skip {quote!{
  #name: self.#name
}} else {
  match ty {
    Type::Path(TypePath { path, .. }) => { ... }
    ...
  }
})
```

DusterTheFirst replied:
I have actually just done that in my commit I'm about  to push

DusterTheFirst replied:
It also lets you use non-Lerp types in structures and just `#[lerp(skip)]` them

DusterTheFirst committed:
clean up derive implementation and update documentation

DusterTheFirst committed:
document lerp(ignore)/lerp(skip)

coriolinus reviewed (approved):
Thanks for all the work you've put into this!
